File: csharp/Basics.cs
namespace csharp
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    using Xunit;
    using static CompositionExtensions;
    using static Collections;

    public class Basics
    {
        [Fact]
        public void Composition()
        {
            IEnumerable<int> yearsOfExperience = new[] { 8, 5, 2, 2, 10, 0 };

            Func<int,bool> isSenior = years => years > 5;

            var elevatedIsSenior = Map(isSenior);

            isSenior(8); // true
            elevatedIsSenior(new[]{ 8, 5, 2}); // [ true, false, true ]


            IEnumerable<bool> seniors1 = yearsOfExperience.Select(IsSenior);
            IEnumerable<bool> seniors2 = yearsOfExperience.Select(isSenior);
            IEnumerable<bool> seniors3 = yearsOfExperience.Select(years => years > 5);

            yearsOfExperience
            .Select(Level)
            .Select(Log)
            .Select(Promote)
            .Select(Log)
            .Select(Years)
            .Select(Log)
            .Select(IsSenior)
            .Select(Log);
            // .ToArray();
        }

        Func<IEnumerable<TSource>, IEnumerable<TResult>> Map<TSource, TResult>(Func<TSource, TResult> func) 
            => source => source.Select(func);

        T Log<T>(T result)
        {
            Console.WriteLine($"LOG {result}");
            return result;
        }
        bool IsSenior(int years) => years > 5;

        string Level(int years) 
            => years switch
            {
                > 10 => "Rockstar",
                > 8  => "Ninja",
                > 5  => "Senior",
                > 3  => "Mid",
                > 0  => "Junior",
                _    => "No idea what I'm doing",
            };

        string Promote(string level) 
            => level switch
            {
                "Ninja"  => "Rockstar",
                "Senior" => "Ninja",
                "Mid"    => "Senior",
                "Junior" => "Mid",
                _        => "Junior",
            };

        int Years(string level) 
            => level switch
            {
                "Rockstar" => 12,
                "Ninja"    => 10,
                "Senior"   => 8,
                "Mid"      => 6,
                "Junior"   => 3,
                _          => 0
            };
    }

    public static class CompositionExtensions {

        public static bool IsEven(this int x) => x % 2 == 0;

        public static string Level(int years) 
            => years switch
            {
                > 10 => "Rockstar",
                > 8  => "Ninja",
                > 5  => "Senior",
                > 3  => "Mid",
                > 0  => "Junior",
                _    => "No idea what I'm doing",
            };

        public static string Promote(string level) 
            => level switch
            {
                "Ninja"  => "Rockstar",
                "Senior" => "Ninja",
                "Mid"    => "Senior",
                "Junior" => "Mid",
                _        => "Junior",
            };

        public static int Years(this string level) 
            => level switch
            {
                "Rockstar" => 12,
                "Ninja"    => 10,
                "Senior"   => 8,
                "Mid"      => 6,
                "Junior"   => 2,
                _          => 0
            };
    }
}

File: fsharp/Basics.fs
module Basics

open Xunit

let add2 x = x + 2
let add5 x = x + 5
let isEven x = x % 2 = 0

let result = add5(add2(1))

let composedFn= add5 >> add2 >> isEven

2 |> add5 |> add2 |> isEven


let level = function
    | x when x > 10 -> "Rockstar"
    | x when x > 8  -> "Ninja"
    | x when x > 5  -> "Senior"
    | x when x > 3  -> "Mid"
    | x when x > 0  -> "Junior"
    | _      -> "No idea what I'm doing"

let promote = function 
    | "Ninja"  -> "Rockstar"
    | "Senior" -> "Ninja"
    | "Mid"    -> "Senior"
    | "Junior" -> "Mid"
    | _        -> "Junior"


let years = function 
    | "Rockstar" -> 12
    | "Ninja"    -> 10
    | "Senior"   -> 8
    | "Mid"      -> 5
    | "Junior"   -> 2
    | _          -> 0 

let isSenior x = x > 5

let yearsOfExperience = [ 8; 5; 2; 2; 10; 0]
let elevatedIsSenior = List.map isSenior

elevatedIsSenior [ 8; 5; 2; 2; 10; 0]

List.map isSenior [ 8; 5; 2; 2; 10; 0]

[ 8; 5; 2; 2; 10; 0] |> List.map isSenior

[ 8; 5; 2; 2; 10; 0] |> List.map (fun x -> x > 5)

let a = 0 |> level |> promote |> years |> isSenior

let b = List.map (level >> promote >> years >> isSenior) [ 8; 5; 2; 2; 10; 0]

let c = [ 8; 5; 2; 2; 10; 0] |> List.map (level >> promote >> years >> isSenior)




